### PR TITLE
igxGrid Exporters

### PR DESCRIPTION
**Pull request by DDavidkov**
#668

Excel, CSV, TSV and TAB exporters

### CODE REVIEW

File: src/services/excel/excel-exporter.ts [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+    private static ZIP_OPTIONS = { compression: "DEFLATE", type: "base64" };
+    private static DATA_URL_PREFIX = "data:application/vnd.openxmlformats-officedocument.spreadsheetml.sheet;base64,";
+
+    private _xlsx: JSZip;
+
+    @Output()
+    public onExportEnded = new EventEmitter<ExcelExportEndedEventArgs>();
+
+    private static PopulateFolder(folder: IExcelFolder, zip: JSZip, worksheetData: WorksheetData): any {
```
Comment (kdinev) [inline src/services/excel/excel-exporter.ts]: Make this method camel-case.

File: src/services/excel/excel-files.ts [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+export class ThemeFile implements IExcelFile {
+    public WriteElement(folder: JSZip, worksheetData: WorksheetData) {
+        folder.file("theme1.xml", ExcelStrings.getTheme());
+    }
+}
+
+export class WorksheetFile implements IExcelFile {
+    private static MIN_WIDTH = 8.34;
+
+    public WriteElement(folder: JSZip, worksheetData: WorksheetData) {
```
Comment (kdinev) [inline src/services/excel/excel-files.ts]: Make it camel-case.

File: src/services/excel/excel-interfaces.ts [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+import {
+    ExcelFileTypes,
+    ExcelFolderTypes
+} from "./excel-enums";
+
+import { WorksheetData } from "./worksheet-data";
+
+export interface IExcelFile {
+    WriteElement(folder: JSZip, data: WorksheetData): void;
```
Comment (kdinev) [inline src/services/excel/excel-interfaces.ts]: Camel-case

File: src/services/excel/excel-interfaces.ts [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+}
+
+export interface IExcelFolder {
+    folderName: string;
+
+    ChildFiles(data: WorksheetData): ExcelFileTypes[];
```
Comment (kdinev) [inline src/services/excel/excel-interfaces.ts]: Camel-case.

**Comment by rkaraivanov** [comment]
@DDavidkov 
I guess we have a dependency on [JSZip](https://github.com/Stuk/jszip) with the exporters. 
This should be added to the changelog/documentation and maybe add it to the `peerDependencies` in the package json

File: src/services/csv/char-separated-value-data.ts [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+        return this._delimiter;
+    }
+
+    public set valueDelimiter(value) {
+        if (value !== undefined && value !== null) {
+            this.setDelimiter(value);
+        }
+    }
+
+    private setDelimiter(value) {
```
Comment (SAndreeva) [inline src/services/csv/char-separated-value-data.ts]: You can move this method down to the rest of the private methods.

File: src/services/csv/csv-exporter-options.ts [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+    constructor(fileName: string, fileType: CsvFileTypes) {
+        super(fileName, IgxCsvExporterOptions.getExtensionFromFileType(fileType));
+        this.setFileType(fileType);
+        this.setDelimiter();
+    }
+
+    private static getExtensionFromFileType(fType: CsvFileTypes) {
+        let extension = "";
+        switch (fType) {
```
Comment (SAndreeva) [inline src/services/csv/csv-exporter-options.ts]: default clause is missing in the switch statement.

File: src/services/csv/csv-exporter-options.ts [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+            if (!this.fileName.endsWith(extension)) {
+                const oldExt = "." + this.fileName.split(".").pop();
+                const newName = this.fileName.replace(oldExt, extension);
+                this._fileExtension = extension;
+                this.fileName = newName;
+            }
+        }
+    }
+
+    private setDelimiter(value?) {
```
Comment (SAndreeva) [inline src/services/csv/csv-exporter-options.ts]: default clause is missing in the switch statement.

File: src/services/excel/jszip-helper.ts [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+`xmlns:x14ac="http://schemas.microsoft.com/office/spreadsheetml/2009/9/ac"><dimension ref="A1"/>` +
+`<sheetViews><sheetView tabSelected="1" workbookViewId="0"></sheetView></sheetViews><sheetFormatPr defaultRowHeight="15" ` +
+`x14ac:dyDescent="0.25"/><sheetData/><pageMargins left="0.7" right="0.7" top="0.75" bottom="0.75" header="0.3" ` +
+`footer="0.3"/></worksheet>`;
+        }
+    }
+
+    public static createExpectedXML(xmlFile: ExcelFileTypes, currentData = "", hasValues = true): any {
+        let resultXml;
+        switch (xmlFile) {
```
Comment (SAndreeva) [inline src/services/excel/jszip-helper.ts]: default clause is missing in the switch statement.

**Pull request merged**